### PR TITLE
Fixed implementation of override_db_setting context manager.

### PR DESCRIPTION
**Pull request by ngnpope**
The `yield` should be after all settings changes have been applied (not that this context manager is used with more than one setting override) and the clean up should happen when the context manager exits, not after the end of the test. This is because the context manager is used multiple times within the same test.

### CODE REVIEW

File: tests/backends/postgresql/test_server_side_cursors.py [hunk 1]
```diff
@@ -31,17 +31,20 @@ def inspect_cursors(self):
...
-                self.addCleanup(operator.delitem, connection.settings_dict, setting)
+                cleanups.append((operator.delitem, settings, name))
+            settings[name] = value
+
+        yield
 
-            connection.settings_dict[setting] = kwargs[setting]
-            yield
+        for operation, *args in reversed(cleanups):
+            operation(*args)
```
Comment (felixxm) [inline tests/backends/postgresql/test_server_side_cursors.py:47]: Can we add a test to shows what was fixed?

Comment (ngnpope) [reply tests/backends/postgresql/test_server_side_cursors.py:47]: Done.

**Comment by ngnpope** [comment]
With the test provided, if the fix is removed, it'll complain on `self.assertEqual(connection.settings_dict["B"], 2)`:

```pytb
Traceback (most recent call last):
  File "/django/source/tests/backends/postgresql/test_server_side_cursors.py", line 64, in test_override_db_setting
    self.assertEqual(connection.settings_dict["B"], 2)
KeyError: 'B'
```

On removing that assertion, it'll then be upset again because of the `yield` being in the wrong place:

```pytb
Traceback (most recent call last):
  File "/django/source/tests/backends/postgresql/test_server_side_cursors.py", line 63, in test_override_db_setting
    with self.override_db_setting(A=1, B=2):
  File "/usr/local/lib/python3.10/contextlib.py", line 146, in __exit__
    raise RuntimeError("generator didn't stop")
RuntimeError: generator didn't stop
```

And, after removing the `B=2`, it'll fail because the clean up isn't happening when the context manager exits:

```pytb
Traceback (most recent call last):
  File "/django/source/tests/backends/postgresql/test_server_side_cursors.py", line 67, in test_override_db_setting
    self.assertNotIn("A", connection.settings_dict)
AssertionError: 'A' unexpectedly found in {'ENGINE': 'django.db.backends.postgresql', 'HOST': 'postgresql-db', 'NAME': 'test_django', 'USER': 'django', 'PASSWORD': 'django', 'ATOMIC_REQUESTS': False, 'AUTOCOMMIT': True, 'CONN_MAX_AGE': 0, 'CONN_HEALTH_CHECKS': False, 'OPTIONS': {}, 'TIME_ZONE': None, 'PORT': '', 'TEST': {'CHARSET': None, 'COLLATION': None, 'MIGRATE': True, 'MIRROR': None, 'NAME': None}, 'A': 1}
```

**Comment by ngnpope** [comment]
Hmm... Hold fire with this. I think there is value in doing a little more here, @felixxm.

**Pull request closed without merging**